### PR TITLE
[Feat] Moonpay

### PR DESCRIPTION
**Pull request by Gamboster**
Required for Moonpay integration with Bitpay wallet: https://github.com/bitpay/bitpay-app/pull/538

### CODE REVIEW

**Review by twiztid89 — APPROVED** [review]

**Review by twiztid89 — APPROVED** [review]

**Review by twiztid89 — APPROVED** [review]

**Review by twiztid89 — APPROVED** [review]

**Review by twiztid89 — APPROVED** [review]

File: packages/bitcore-wallet-service/src/lib/server.ts [hunk 1]
```diff
@@ -4534,6 +4534,191 @@ export class WalletService {
     return cb(null, data);
   }
 
+  moonpayGetKeys(req) {
```
Comment (kajoseph) [inline packages/bitcore-wallet-service/src/lib/server.ts:4537]: Let's go ahead and make this a `private` function

**Review by kajoseph — CHANGES_REQUESTED** [review]

File: packages/bitcore-wallet-service/src/lib/server.ts [hunk 2]
```diff
@@ -4534,6 +4534,191 @@ export class WalletService {
...
+            return resolve(data.body ? data.body : data);
+          }
+        }
+      );
+
+    });
+  }
+
+  moonpayGetSignedPaymentUrl(req): Promise<{urlWithSignature: string}> {
+    return new Promise((resolve, reject) => {
```
Comment (kajoseph) [inline packages/bitcore-wallet-service/src/lib/server.ts:4607]: Some of these functions like this one don't need to be a promise, right?

**Review by kajoseph — CHANGES_REQUESTED** [review]

File: packages/bitcore-wallet-service/src/lib/expressapp.ts [hunk 3]
```diff
@@ -1498,6 +1498,54 @@ export class ExpressApp {
...
 
+    router.post('/v1/service/moonpay/quote', async(req, res) => {
+      getServerWithAuth(req, res, server => {
+        server
+          .moonpayGetQuote(req)
+          .then(response => {
+            res.json(response);
+          })
+          .catch(err => {
+            if (err) return returnError(err, res, req);
```
[the remark this answers is not in the document]
Comment (Gamboster) [reply packages/bitcore-wallet-service/src/lib/expressapp.ts]: Done

File: packages/bitcore-wallet-service/src/lib/expressapp.ts [hunk 4]
```diff
@@ -1498,6 +1498,54 @@ export class ExpressApp {
...
+
+    router.post('/v1/service/moonpay/signedPaymentUrl', async(req, res) => {
+      getServerWithAuth(req, res, server => {
+        server
+          .moonpayGetSignedPaymentUrl(req)
+          .then(response => {
+            res.json(response);
+          })
+          .catch(err => {
+            if (err) return returnError(err, res, req);
```
[the remark this answers is not in the document]
Comment (Gamboster) [reply packages/bitcore-wallet-service/src/lib/expressapp.ts]: Done

File: packages/bitcore-wallet-service/src/lib/server.ts (hunk 1 above)
Comment (Gamboster) [reply packages/bitcore-wallet-service/src/lib/server.ts]: Done

File: packages/bitcore-wallet-service/src/lib/expressapp.ts [hunk 5]
```diff
@@ -1498,6 +1498,54 @@ export class ExpressApp {
       }
     });
 
+    router.post('/v1/service/moonpay/quote', async(req, res) => {
+      getServerWithAuth(req, res, server => {
+        server
+          .moonpayGetQuote(req)
+          .then(response => {
```
[the remark this answers is not in the document]
Comment (Gamboster) [reply packages/bitcore-wallet-service/src/lib/expressapp.ts]: Done

File: packages/bitcore-wallet-service/src/lib/expressapp.ts [hunk 6]
```diff
@@ -1498,6 +1498,54 @@ export class ExpressApp {
...
+            if (err) return returnError(err, res, req);
+          });
+      });
+    });
+
+    router.post('/v1/service/moonpay/signedPaymentUrl', async(req, res) => {
+      getServerWithAuth(req, res, server => {
+        server
+          .moonpayGetSignedPaymentUrl(req)
+          .then(response => {
```
[the remark this answers is not in the document]
Comment (Gamboster) [reply packages/bitcore-wallet-service/src/lib/expressapp.ts]: Done

File: packages/bitcore-wallet-service/src/lib/server.ts (hunk 2 above)
Comment (Gamboster) [reply packages/bitcore-wallet-service/src/lib/server.ts]: Promise removed. Ty

File: packages/bitcore-wallet-service/test/integration/server.js [hunk 7]
```diff
@@ -10624,6 +10624,283 @@ describe('Wallet service', function() {
...
+            env: 'sandbox',
+            currencyAbbreviation: 'btc',
+            baseCurrencyAmount: 50,
+            extraFeePercentage: 5,
+            baseCurrencyCode: 'usd'
+          }
+        }
+      });
+  
+      it('should work properly if req is OK', () => {
```
Comment (kajoseph) [inline packages/bitcore-wallet-service/test/integration/server.js:10686]: These tests need a callback

**Review by kajoseph — CHANGES_REQUESTED** [review]
One last thing

File: packages/bitcore-wallet-service/test/integration/server.js (hunk 7 above)
Comment (Gamboster) [reply packages/bitcore-wallet-service/test/integration/server.js]: Done, ty

**Review by kajoseph — APPROVED** [review]

**Pull request merged**